File: src/main/java/org/proxydroid/ConnectivityBroadcastReceiver.java
/* proxydroid - Global / Individual Proxy App for Android
 * Copyright (C) 2011 Max Lv <max.c.lv@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 * 
 *                            ___====-_  _-====___
 *                      _--^^^#####//      \\#####^^^--_
 *                   _-^##########// (    ) \\##########^-_
 *                  -############//  |\^^/|  \\############-
 *                _/############//   (@::@)   \\############\_
 *               /#############((     \\//     ))#############\
 *              -###############\\    (oo)    //###############-
 *             -#################\\  / VV \  //#################-
 *            -###################\\/      \//###################-
 *           _#/|##########/\######(   /\   )######/\##########|\#_
 *           |/ |#/\#/\#/\/  \#/\##\  |  |  /##/\#/  \/\#/\#/\#| \|
 *           `  |/  V  V  `   V  \#\| |  | |/#/  V   '  V  V  \|  '
 *              `   `  `      `   / | |  | | \   '      '  '   '
 *                               (  | |  | |  )
 *                              __\ | |  | | /__
 *                             (vvv(VVV)(VVV)vvv)
 *
 *                              HERE BE DRAGONS
 *
 */

package org.proxydroid;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.util.Log;
import com.ksmaze.android.preference.ListPreferenceMultiSelect;
import org.proxydroid.utils.Constraints;
import org.proxydroid.utils.Utils;

public class ConnectivityBroadcastReceiver extends BroadcastReceiver {

  private Handler mHandler = new Handler();

  private static final String TAG = "ConnectivityBroadcastReceiver";

  @Override
  public void onReceive(final Context context, final Intent intent) {

    if (Utils.isConnecting()) return;

    if (!intent.getAction().equals(ConnectivityManager.CONNECTIVITY_ACTION)) return;

    mHandler.post(new Runnable() {
      @Override
      public void run() {

        // only switching profiles when needed
        ConnectivityManager manager = (ConnectivityManager) context
            .getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = manager.getActiveNetworkInfo();

        if (networkInfo != null) {
          if (networkInfo.getState() == NetworkInfo.State.CONNECTING
              || networkInfo.getState() == NetworkInfo.State.DISCONNECTING
              || networkInfo.getState() == NetworkInfo.State.UNKNOWN)
            return;
        } else {
          if (!Utils.isWorking()) return;
        }

        SharedPreferences settings = PreferenceManager
            .getDefaultSharedPreferences(context);
        Profile mProfile = new Profile();
        mProfile.getProfile(settings);

        // Store current settings first
        String oldProfile = settings.getString("profile", "1");
        Editor ed = settings.edit();
        ed.putString(oldProfile, mProfile.toString());
        ed.commit();

        // Load all profiles
        String[] profileValues = settings.getString("profileValues", "").split(
            "\\|");
        String curSSID = null;
        String lastSSID = settings.getString("lastSSID", "-1");
        boolean autoConnect = false;

        // Test on each profile
        for (String profile : profileValues) {
          String profileString = settings.getString(profile, "");
          mProfile.decodeJson(profileString);
          curSSID = onlineSSID(context, mProfile.getSsid());
          if (mProfile.isAutoConnect() && curSSID != null) {
            // Enable auto connect
            autoConnect = true;

            // XXX: Switch profile first
            ed = settings.edit();
            ed.putString("profile", profile);
            ed.commit();

            // Then switch profile values
            mProfile.setProfile(settings);
            break;
          }
        }

        if (networkInfo == null) {
          if (!lastSSID.equals(Constraints.ONLY_3G)
              && !lastSSID.equals(Constraints.WIFI_AND_3G)
              && !lastSSID.equals(Constraints.ONLY_WIFI)) {
            if (Utils.isWorking()) {
              context.stopService(new Intent(context,
                  ProxyDroidService.class));
            }
          }
        } else {
          // no network available now
          if (networkInfo.getState() != NetworkInfo.State.CONNECTED)
            return;

          if (networkInfo.getType() == ConnectivityManager.TYPE_WIFI) {
            // if no last SSID, should give up here
            if (!lastSSID.equals("-1")) {
              // get WIFI info
              WifiManager wm = (WifiManager) context
                  .getSystemService(Context.WIFI_SERVICE);
              WifiInfo wInfo = wm.getConnectionInfo();
              if (wInfo != null) {
                // compare with the current SSID
                String current = wInfo.getSSID();
                if (current != null) current = current.replace("\"", "");
                if (current != null && !current.equals(lastSSID)) {
                  // need to switch profile, so stop service first
                  if (Utils.isWorking())
                    context.stopService(new Intent(context,
                        ProxyDroidService.class));
                }
              }
            }
          } else {
            // still satisfy the last trigger
            if (!lastSSID.equals(Constraints.ONLY_3G)
                && !lastSSID.equals(Constraints.WIFI_AND_3G)) {
              if (Utils.isWorking())
                context.stopService(new Intent(context,
                    ProxyDroidService.class));
            }
          }
        }

        if (autoConnect) {
          if (!Utils.isWorking()) {
            ProxyDroidReceiver pdr = new ProxyDroidReceiver();
            ed = settings.edit();
            ed.putString("lastSSID", curSSID);
            ed.commit();
            Utils.setConnecting(true);
            pdr.onReceive(context, intent);
          }
        }
      }
    });
  }

  public String onlineSSID(Context context, String ssid) {
    String ssids[] = ListPreferenceMultiSelect.parseStoredValue(ssid);
    if (ssids == null)
      return null;
    if (ssids.length < 1)
      return null;
    ConnectivityManager manager = (ConnectivityManager) context
        .getSystemService(Context.CONNECTIVITY_SERVICE);
    NetworkInfo networkInfo = manager.getActiveNetworkInfo();
    if (networkInfo == null)
      return null;
    if (networkInfo.getType() != ConnectivityManager.TYPE_WIFI) {
      for (String item : ssids) {
        if (Constraints.WIFI_AND_3G.equals(item))
          return item;
        if (Constraints.ONLY_3G.equals(item))
          return item;
      }
      return null;
    }
    WifiManager wm = (WifiManager) context
        .getSystemService(Context.WIFI_SERVICE);
    WifiInfo wInfo = wm.getConnectionInfo();
    if (wInfo == null || wInfo.getSSID() == null)
      return null;
    String current = wInfo.getSSID();
    if (current == null || "".equals(current))
      return null;
    current = current.replace("\"", "");
    for (String item : ssids) {
      if (Constraints.WIFI_AND_3G.equals(item))
        return item;
      if (Constraints.ONLY_WIFI.equals(item))
        return item;
      if (current.equals(item))
        return item;
    }
    return null;
  }

}

File: app/src/main/java/org/proxydroid/ConnectivityBroadcastReceiver.kt
/* proxydroid - Global / Individual Proxy App for Android
 * Copyright (C) 2011 Max Lv <max.c.lv@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

package org.proxydroid

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.wifi.WifiManager
import android.preference.PreferenceManager
import com.ksmaze.android.preference.ListPreferenceMultiSelect
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.proxydroid.utils.Constraints
import org.proxydroid.utils.ProxyController
import org.proxydroid.utils.Utils

class ConnectivityBroadcastReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (Utils.isConnecting()) return
        if (intent.action != ConnectivityManager.CONNECTIVITY_ACTION) return

        // BroadcastReceiver runs on the main thread; offload to IO with goAsync().
        val pending = goAsync()
        scope.launch {
            try {
                handle(context.applicationContext, intent)
            } finally {
                pending.finish()
            }
        }
    }

    private fun handle(context: Context, intent: Intent) {
        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = manager.activeNetworkInfo

        if (networkInfo != null) {
            val state = networkInfo.state
            if (state == NetworkInfo.State.CONNECTING ||
                state == NetworkInfo.State.DISCONNECTING ||
                state == NetworkInfo.State.UNKNOWN
            ) return
        } else if (!Utils.isWorking()) {
            return
        }

        val settings = PreferenceManager.getDefaultSharedPreferences(context)
        val profile = Profile().apply { getProfile(settings) }

        val oldProfile = settings.getString("profile", "1") ?: "1"
        settings.edit().putString(oldProfile, profile.toString()).apply()

        val profileValues = settings.getString("profileValues", "")
            ?.split("|")?.filter { it.isNotEmpty() }
            .orEmpty()
        var curSSID: String? = null
        val lastSSID = settings.getString("lastSSID", "-1") ?: "-1"
        var autoConnect = false

        for (profileId in profileValues) {
            val profileString = settings.getString(profileId, "") ?: ""
            profile.decodeJson(profileString)
            curSSID = onlineSSID(context, profile.ssid, profile.excludedSsid)
            if (profile.isAutoConnect && curSSID != null) {
                autoConnect = true
                settings.edit().putString("profile", profileId).apply()
                profile.setProfile(settings)
                break
            }
        }

        if (networkInfo == null) {
            if (lastSSID != Constraints.ONLY_3G &&
                lastSSID != Constraints.WIFI_AND_3G &&
                lastSSID != Constraints.ONLY_WIFI &&
                Utils.isWorking()
            ) {
                ProxyController.stop(context)
            }
        } else {
            if (networkInfo.state != NetworkInfo.State.CONNECTED) return

            if (networkInfo.type == ConnectivityManager.TYPE_WIFI) {
                if (lastSSID != "-1") {
                    val wm = context.applicationContext
                        .getSystemService(Context.WIFI_SERVICE) as WifiManager
                    val current = wm.connectionInfo?.ssid?.replace("\"", "")
                    if (current != null && current != lastSSID && Utils.isWorking()) {
                        ProxyController.stop(context)
                    }
                }
            } else if (lastSSID != Constraints.ONLY_3G &&
                lastSSID != Constraints.WIFI_AND_3G &&
                Utils.isWorking()
            ) {
                ProxyController.stop(context)
            }
        }

        if (autoConnect && !Utils.isWorking()) {
            settings.edit().putString("lastSSID", curSSID).apply()
            Utils.setConnecting(true)
            ProxyDroidReceiver().onReceive(context, intent)
        }
    }

    private fun onlineSSID(context: Context, ssid: String, excludedSsid: String): String? {
        val ssids = ListPreferenceMultiSelect.parseStoredValue(ssid) ?: return null
        val excludedSsids = ListPreferenceMultiSelect.parseStoredValue(excludedSsid)

        if (ssids.isEmpty()) return null

        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = manager.activeNetworkInfo ?: return null

        if (networkInfo.type != ConnectivityManager.TYPE_WIFI) {
            for (item in ssids) {
                if (Constraints.WIFI_AND_3G == item) return item
                if (Constraints.ONLY_3G == item) return item
            }
            return null
        }

        val wm = context.applicationContext
            .getSystemService(Context.WIFI_SERVICE) as WifiManager
        val current = wm.connectionInfo?.ssid?.takeIf { it.isNotEmpty() }
            ?.replace("\"", "") ?: return null

        if (excludedSsids != null) {
            for (item in excludedSsids) {
                if (current == item) return null
            }
        }

        for (item in ssids) {
            if (Constraints.WIFI_AND_3G == item) return item
            if (Constraints.ONLY_WIFI == item) return item
            if (current == item) return item
        }
        return null
    }

    companion object {
        private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    }
}
